proposals/online-migrations-poc-and-proposal.md

```diff
@@ -0,0 +1,183 @@
+`cleanup_migrations_status_table`. This is mostly boilerplate thanks to
+`goose`.
+
+#### Table versions
+
+Using different table versions is easy. The intuition of using a `const` to set
+the table name in a place works fine, just a bit boilerplate-y to build
+queries, which shouldn't be a problem with the right abstraction.
+
+### Complexities and limitations
+
+The process described works for a simple table, things get more complex
+with foreign keys and relationships.
+
+**Parent tables**
+
+A major concern with this approach is with "parent" tables, tables that are
+referenced by other tables and have foreign key constraints on them.
+
+Updating a foreign key to point to the new table requires an `ALTER` statement,
+we've to do the whole versioning procedure for any "child" table referencing a
+"parent" table being modified.
+
+This problem is also recursive, because a "child" table can itself be
+referenced by other tables (for example the `users` table)
+
+I did write an SQL migration simulating this scenario, but it's long enough to
+not be included in this document.
+
+Having said that, we can mitigate this problem with two considerations:
+
+1. As mentioned in the original document, we can use `ALTER` directly on
+   eventually consistent tables.
+2. If a table is small enough (we should test before rolling the change)
+   `ALTER`s aren't locking.
+
+**Child tables**
+
+The simplest case is a "child" table that references another table and has a
+foreign key constraint. Certain aspects require special care:
+
+1. Explicitly create the foreign key constraints on the new table
+2. While both tables are "live," to `DELETE` a record from the parent table we
+   first have to delete all references from both child tables. This is covered
+   with triggers, but is an area that we should load test and inspect for write
+   races.
+
+**Races and duplicates**
+
+Another general concern is with data races and the fact we use `AUTO_INCREMENT`
+IDs. Can we have conflicts if an insert is performed simultaneously in
+`table_n` and `table_n+1`? is this realistic?
+
+## Alternative proposal
+
+Another way to tackle this problem could be with the combination of:
+
+1. Non-breaking DB changes between two immediate Fleet versions, so both Fleet
+   versions can use the same database.
+2. [MySQL online DDL][mysql-ddl], which lets you to write non-locking alter
+   statements via `ALTER ... LOCK=NONE`. This functionality is not available
+   for certain operations (notably changing a column data type) but they're
+   very few, all have a workaround, and per `1` we don't want immediate
+   breaking changes anyway.
+
+Common scenarios (based on our latest 10 migrations):
+
+- Adding a new column: as long as isn't `AUTO_INCREMENT`, `LOCK=NONE` is
+  allowed
+- Renaming a column: don’t rename, instead
+  - `vN`: non breaking, add a new column with `LOCK=NONE`, allow `NULL` values.
+  - `vN+1: breaking, remove the old column, add `NOT NULL` constraints if
```

Comment:
Nit: You missed the closing quote here.

Reply:
Any ideas how we would make sure to drop the column at the right point in time?

Reply:
fixed, thanks!

> Any ideas how we would make sure to drop the column at the right point in time?

for what I have been investigating, timing is not crucial as long as you keep the house in order, it shouldn't matter that much if we remove the table in `vN+2`, as long as we remove it.

in regards to enforcing and making sure we eventually remove the table, I'm not sure 🤔. Ideally there's some kind of automation as I don't trust us Humans in regards to consistency.

Reply:
If the PR has an ddl change, then a CI could start failing once a given tag is created. Eg:

If we are in version 4.15, we could be forced to add a line on a file that reads:

```
fleet-4.15.0 remove column blah
```

So if there's a new tag for fleet, CI could fail until there's a PR that removed that line and adds the migration.

Sounds convoluted, so I don't know if it's the best, but the best I can come up with right now at least.

Reply:
I don't find it that convoluted, and I like the simplicity of the idea behind it, but agreed that we might come up with something else.

Reply:
another way, inspired by [this](https://github.com/searls/todo_or_die) ruby gem I found randomly could be to add a helper function:

```go
func DoBefore(msg, fleetVersion string) {
  if isCi  && fleetVersion > currentFleetVersion {
    panic("you should " + msg + " before " + fleetVersion)
  }
}

// somewhere in the code:
DoBefore("drop x column in hosts table", "4.18.0")
```

we could also make it a compile time error if we get creative and hacky, based on [this](https://groups.google.com/g/golang-nuts/c/yT5DD0ZDI9c)